File: tests/test_api_agent.py
"""API integration test for full system routing via router agent."""
from __future__ import annotations

import asyncio
import os
from pathlib import Path
import unittest

import pandas as pd
import uuid_utils

from dotenv import load_dotenv, find_dotenv
from agents import Runner
from agents.extensions.memory import AsyncSQLiteSession

from neuro_seller.router import router_agent


# Load .env using find_dotenv (more robust in monorepos / nested runs)
load_dotenv(find_dotenv())

ROOT = Path(__file__).resolve().parents[1]
TESTS_DIR = ROOT / "tests"


class TestApiSystemFromExcel(unittest.TestCase):
    """Integration test for full routing system (router + tools)."""

    @classmethod
    def setUpClass(cls) -> None:
        if os.getenv("RUN_API_TESTS") != "1":
            raise unittest.SkipTest(
                "Set RUN_API_TESTS=1 to run API integration tests."
            )

        if not os.getenv("OPENAI_API_KEY"):
            raise RuntimeError("OPENAI_API_KEY is not configured.")

    def _extract_toolspan(self, result) -> str:
        """
        Extract tool/agent used by router (Agent-as-Tool).
        Fallback: 'unknown'
        """
        try:
            for item in getattr(result, "new_items", []):

                # direct name (newer SDK)
                name = getattr(item, "name", None)
                if name:
                    return str(name)

                # raw item fallback
                raw = getattr(item, "raw_item", None)
                if raw:
                    name = getattr(raw, "name", None)
                    if name:
                        return str(name)

        except Exception:
            pass

        return "unknown"

    async def _run_once(self, text: str) -> tuple[str, str, str]:
        session_id = str(uuid_utils.uuid7())

        session = AsyncSQLiteSession(
            session_id=session_id,
            db_path="users.db",
        )

        try:
            result = await Runner.run(
                starting_agent=router_agent,
                input=text,
                session=session,
            )

            response = str(result.final_output)
            toolspan = self._extract_toolspan(result)

            return response, toolspan, session_id

        finally:
            await session.close()

    def _process_file(self, source_file: str, result_file: str) -> None:
        input_path = TESTS_DIR / source_file

        if not input_path.exists():
            self.fail(f"Input file not found: {input_path}")

        df = pd.read_excel(input_path)

        if "request" not in df.columns:
            self.fail(f"Column 'request' not found in {source_file}")

        limit = os.getenv("API_TEST_LIMIT")
        if limit:
            df = df.head(int(limit))

        responses: list[str] = []
        toolspans: list[str] = []
        session_ids: list[str] = []

        for request in df["request"].fillna("").astype(str):
            response, toolspan, session_id = asyncio.run(
                self._run_once(request)
            )

            responses.append(response)
            toolspans.append(toolspan)
            session_ids.append(session_id)

        output_df = pd.DataFrame(
            {
                "row_id": range(1, len(df) + 1),
                "request": df["request"],
                "response": responses,
                "toolspan": toolspans,
                "session_id": session_ids,
            }
        )

        output_path = TESTS_DIR / result_file
        output_df.to_excel(output_path, index=False)

        print(f"\nSaved: {output_path}")

    def test_process_consult(self) -> None:
        self._process_file(
            source_file="for_test_consult.xlsx",
            result_file="result_consult.xlsx",
        )

    def test_process_goodbye(self) -> None:
        self._process_file(
            source_file="for_test_goodbye.xlsx",
            result_file="result_goodbye.xlsx",
        )


if __name__ == "__main__":
    unittest.main()

File: tests/test-api-agent.ts
import 'dotenv/config';
import assert from 'node:assert/strict';
import { existsSync } from 'node:fs';
import { test } from 'node:test';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';

import XLSX from 'xlsx';

import { runRouterOnce } from '../neuro_seller/router.js';

type ApiRow = {
  request?: string;
};

type ResultRow = {
  row_id: number;
  request: string;
  response: string;
  toolspan: string;
  session_id: string;
};

const root = join(dirname(fileURLToPath(import.meta.url)), '..');
const testsDir = join(root, 'tests');

function extractToolspan(result: { newItems?: unknown[] }): string {
  for (const item of result.newItems ?? []) {
    if (typeof item !== 'object' || item === null) {
      continue;
    }

    const candidate = item as Record<string, unknown>;
    const directName = candidate.name;
    if (typeof directName === 'string' && directName.length > 0) {
      return directName;
    }

    const rawItem = candidate.rawItem;
    if (typeof rawItem === 'object' && rawItem !== null) {
      const rawName = (rawItem as Record<string, unknown>).name;
      if (typeof rawName === 'string' && rawName.length > 0) {
        return rawName;
      }
    }
  }

  return 'unknown';
}

async function processFile(sourceFile: string, resultFile: string): Promise<void> {
  const inputPath = join(testsDir, sourceFile);
  assert.ok(existsSync(inputPath), `Input file not found: ${inputPath}`);

  const workbook = XLSX.readFile(inputPath);
  const firstSheetName = workbook.SheetNames[0];
  assert.ok(firstSheetName, `No worksheets found in ${sourceFile}`);

  const worksheet = workbook.Sheets[firstSheetName];
  const inputRows = XLSX.utils.sheet_to_json<ApiRow>(worksheet, { defval: '' });

  assert.ok(
    inputRows.length === 0 || Object.hasOwn(inputRows[0], 'request'),
    `Column 'request' not found in ${sourceFile}`,
  );

  const limit = process.env.API_TEST_LIMIT ? Number(process.env.API_TEST_LIMIT) : undefined;
  const rows = Number.isInteger(limit) && limit !== undefined ? inputRows.slice(0, limit) : inputRows;

  const resultRows: ResultRow[] = [];

  for (const [index, row] of rows.entries()) {
    const request = String(row.request ?? '');
    const { response, sessionId, result } = await runRouterOnce(request);

    resultRows.push({
      row_id: index + 1,
      request,
      response,
      toolspan: extractToolspan(result),
      session_id: sessionId,
    });
  }

  const outputWorkbook = XLSX.utils.book_new();
  const outputWorksheet = XLSX.utils.json_to_sheet(resultRows);
  XLSX.utils.book_append_sheet(outputWorkbook, outputWorksheet, 'results');
  XLSX.writeFile(outputWorkbook, join(testsDir, resultFile));
}

const runApiTests = process.env.RUN_API_TESTS === '1';
const apiTest = runApiTests ? test : test.skip;

apiTest('process consult Excel requests through router', async () => {
  assert.ok(process.env.OPENAI_API_KEY, 'OPENAI_API_KEY is not configured.');
  await processFile('for_test_consult.xlsx', 'result_consult.xlsx');
});

apiTest('process goodbye Excel requests through router', async () => {
  assert.ok(process.env.OPENAI_API_KEY, 'OPENAI_API_KEY is not configured.');
  await processFile('for_test_goodbye.xlsx', 'result_goodbye.xlsx');
});
